src/map_votes.cpp

```diff
@@ -0,0 +1,543 @@
+	}
+	vecMapNames.Sort(OrderStringsLexicographically);
+	FOR_EACH_VEC(vecMapNames, i)
+	{
+		ClientPrint(player, HUD_PRINTCONSOLE, "- %s", vecMapNames[i]);
+	}
+}
+
+CON_COMMAND_CHAT_FLAGS(nomlist, "List the list of nominations", ADMFLAG_NONE)
+{
+	ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "Current nominations:");
+	for (int i = 0; i < g_pMapVoteSystem->GetMapListSize(); i++) {
+		if (!g_pMapVoteSystem->IsMapIndexEnabled(i)) continue;
+		int iNumNominations = g_pMapVoteSystem->GetTotalNominations(i);
+		if (iNumNominations == 0) continue;
+		const char* sMapName = g_pMapVoteSystem->GetMapName(i);
+		ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "- %s (%d times)\n", sMapName, iNumNominations);
+	}
+}
+
+CON_COMMAND_CHAT_FLAGS(mapcooldowns, "List the maps currently in cooldown", ADMFLAG_NONE)
+{
+	ClientPrint(player, HUD_PRINTTALK, CHAT_PREFIX "The list of maps in cooldown will be shown in console.");
+	ClientPrint(player, HUD_PRINTCONSOLE, "The list of maps in cooldown is:");
+	int iMapsInCooldown = g_pMapVoteSystem->GetMapsInCooldown();
+	for (int i = iMapsInCooldown - 1; i >= 0; i--) {
+		int iMapIndex = g_pMapVoteSystem->GetCooldownMap(i);
+		const char* sMapName = g_pMapVoteSystem->GetMapName(iMapIndex);
+		ClientPrint(player, HUD_PRINTCONSOLE, "- %s (%d maps ago)", sMapName, iMapsInCooldown - i);
+	}
+}
+
+GAME_EVENT_F(cs_win_panel_match)
+{
+	g_pMapVoteSystem->StartVote();
+}
+
+GAME_EVENT_F(endmatch_mapvote_selecting_map)
+{
+	g_pMapVoteSystem->FinishVote();
+}
+
+bool CMapVoteSystem::IsMapIndexEnabled(int iMapIndex)
+{
+	if (iMapIndex >= m_vecMapList.Count() || iMapIndex < 0) return false;
+	if (m_vecLastPlayedMapIndexes.HasElement(iMapIndex)) return false;
+	return m_vecMapList[iMapIndex].IsEnabled();
+}
+
+void CMapVoteSystem::StartVote() 
+{
+	m_bIsVoteOngoing = true;
+
+	// If we are forcing a map, just set all vote options to that map
+	if (m_iForcedNextMapIndex != -1) {
+		for (int i = 0; i < 10; i++) {
+			g_pGameRules->m_nEndMatchMapGroupVoteOptions[i] = m_iForcedNextMapIndex;
+			return;
+		}
+	}
+
+	// Seed the randomness for the event
+	int iRandomSeed = (int)gpGlobals->curtime;
+	srand(iRandomSeed);
+	m_iRandomWinnerShift = rand();
+
+	// Reset the player vote counts as the vote just started
+	for (int i = 0; i < gpGlobals->maxClients; i++) {
+		m_arrPlayerVotes[i] = -1;
+	}
+
+	// Select random maps not in cooldown, not disabled, and not nominated
+	CUtlVector<int> vecPossibleMaps;
+	CUtlVector<int> vecIncludedMaps;
+	GetNominatedMapsForVote(vecIncludedMaps);
+	for (int i = 0; i < m_vecMapList.Count(); i++) {
+		if (!IsMapIndexEnabled(i)) continue;
+		if (vecIncludedMaps.HasElement(i)) continue;
+		vecPossibleMaps.AddToTail(i);
+	}
+
+	// Print all available maps out to console
+	FOR_EACH_VEC(vecPossibleMaps, i)
+	{
+		int iPossibleMapIndex = vecPossibleMaps[i];
+		Msg("The %d-th possible map index %d is %s\n", i, iPossibleMapIndex, m_vecMapList[iPossibleMapIndex].GetName());
```

Comment:
Use `Message()` for all server console logs, it's the same as `Msg` but adds a prefix and color.